gallery generation broken on cpython master

Have not had time to dig into why (or have a minimal example), but seeing things like

```
jupiter@11:39  ➤  make html
Running Sphinx v1.8.4
Adding copy buttons to code blocks...
[autosummary] generating autosummary for: api/_as_gen/matplotlib.animation.AVConvBase.rst, api/_as_gen/matplotlib.animation.AVConvFileWriter.rst, api/_as_gen/matplotlib.animation.AVConvWriter.rst, api/_as_gen/matplotlib.animation.AbstractMovieWriter.rst, api/_as_gen/matplotlib.animation.Animation.rst, api/_as_gen/matplotlib.animation.ArtistAnimation.rst, api/_as_gen/matplotlib.animation.FFMpegBase.rst, api/_as_gen/matplotlib.animation.FFMpegFileWriter.rst, api/_as_gen/matplotlib.animation.FFMpegWriter.rst, api/_as_gen/matplotlib.animation.FileMovieWriter.rst, ..., users/prev_whats_new/whats_new_1.2.2.rst, users/prev_whats_new/whats_new_1.2.rst, users/prev_whats_new/whats_new_1.3.rst, users/prev_whats_new/whats_new_1.4.rst, users/prev_whats_new/whats_new_1.5.rst, users/prev_whats_new/whats_new_2.0.0.rst, users/prev_whats_new/whats_new_2.1.0.rst, users/prev_whats_new/whats_new_2.2.rst, users/shell.rst, users/whats_new.rst
loading intersphinx inventory from https://docs.python.org/3/objects.inv...
loading intersphinx inventory from https://docs.scipy.org/doc/numpy/objects.inv...
loading intersphinx inventory from https://docs.scipy.org/doc/scipy/reference/objects.inv...
loading intersphinx inventory from https://pandas.pydata.org/pandas-docs/stable/objects.inv...
loading intersphinx inventory from https://pillow.readthedocs.io/en/stable/objects.inv...
loading intersphinx inventory from https://matplotlib.org/cycler/objects.inv...
generating gallery...
generating gallery for gallery/lines_bars_and_markers... [  2%] bar_stacked.py                                                                                                                                                                                                                       
Warning, treated as error:
/home/tcaswell/source/p/matplotlib/examples/lines_bars_and_markers/bar_stacked.py failed to execute correctly: Traceback (most recent call last):
  File "/home/tcaswell/source/p/matplotlib/examples/lines_bars_and_markers/bar_stacked.py", line 1
    =================
    ^
SyntaxError: invalid syntax

make: *** [Makefile:30: html] Error 2

```
when trying to build the Matplotlib docs.

This makes me think that the module docstring is not being correctly identified as such.  Similar things have happened before when details of the AST changed.

--- a/.travis.yml
+++ b/.travis.yml
@@ -48,8 +48,6 @@ matrix:
             libblas-dev
             liblapack-dev
             gfortran
-  allow_failures:
-    - python: "nightly"
 
 before_install:
   # Make sure that things work even if the locale is set to C (which

--- a/continuous_integration/travis/test_script.sh
+++ b/continuous_integration/travis/test_script.sh
@@ -8,7 +8,7 @@ set -e
 
 pytest sphinx_gallery
 cd doc
-if [ "$DISTRIB" != "minimal" ] || [ "$PYTHON_VERSION" == "nightly" ]; then
+if [ "$DISTRIB" != "minimal" ] && [ "$PYTHON_VERSION" != "nightly" ]; then
     make html-noplot
     make html -j 2
 fi

--- a/setup.cfg
+++ b/setup.cfg
@@ -1,5 +1,5 @@
 [tool:pytest]
-addopts = --cov-report term-missing --cov=sphinx_gallery --durations=5 -ra
+addopts = --cov-report= --cov=sphinx_gallery --durations=5 -ra
 python_files = tests/*.py
 norecursedirs = build _build auto_examples gen_modules
                 sphinx_gallery/tests/tinybuild

--- a/sphinx_gallery/py_source_parser.py
+++ b/sphinx_gallery/py_source_parser.py
@@ -8,7 +8,11 @@ Parser for python source files
 
 from __future__ import division, absolute_import, print_function
 import ast
+from distutils.version import LooseVersion
+from io import BytesIO
 import re
+import sys
+import tokenize
 from textwrap import dedent
 
 SYNTAX_ERROR_DOCSTRING = """
@@ -73,49 +77,43 @@ def get_docstring_and_rest(filename):
     if not isinstance(node, ast.Module):
         raise TypeError("This function only supports modules. "
                         "You provided {0}".format(node.__class__.__name__))
-    try:
-        # in python 3.7 module knows it's docstring
-        # everything else will raise an attribute error
-        docstring = node.docstring
-
-        import tokenize
-        from io import BytesIO
-        ts = tokenize.tokenize(BytesIO(content).readline)
-        ds_lines = 0
-        # find the first string according to the tokenizer and get
-        # it's end row
-        for tk in ts:
-            if tk.exact_type == 3:
-                ds_lines, _ = tk.end
-                break
-        # grab the rest of the file
-        rest = '\n'.join(content.split('\n')[ds_lines:])
-        lineno = ds_lines + 1
-
-    except AttributeError:
-        # this block can be removed when python 3.6 support is dropped
-        if node.body and isinstance(node.body[0], ast.Expr) and \
-           isinstance(node.body[0].value, ast.Str):
-            docstring_node = node.body[0]
-            docstring = docstring_node.value.s
-            # python2.7: Code was read in bytes needs decoding to utf-8
-            # unless future unicode_literals is imported in source which
-            # make ast output unicode strings
-            if hasattr(docstring, 'decode') and not isinstance(docstring, unicode):
-                docstring = docstring.decode('utf-8')
-            lineno = docstring_node.lineno  # The last line of the string.
-            # This get the content of the file after the docstring last line
-            # Note: 'maxsplit' argument is not a keyword argument in python2
-            rest = content.split('\n', lineno)[-1]
-            lineno += 1
-        else:
-            docstring, rest = '', ''
-
-    if not docstring:
+    if not (node.body and isinstance(node.body[0], ast.Expr) and
+            isinstance(node.body[0].value, ast.Str)):
         raise ValueError(('Could not find docstring in file "{0}". '
                           'A docstring is required by sphinx-gallery '
                           'unless the file is ignored by "ignore_pattern"')
                          .format(filename))
+
+    if LooseVersion(sys.version) >= LooseVersion('3.7'):
+        docstring = ast.get_docstring(node)
+        assert docstring is not None  # should be guaranteed above
+        # This is just for backward compat
+        if len(node.body[0].value.s) and node.body[0].value.s[0] == '\n':
+            # just for strict backward compat here
+            docstring = '\n' + docstring
+        ts = tokenize.tokenize(BytesIO(content.encode()).readline)
+        # find the first string according to the tokenizer and get its end row
+        for tk in ts:
+            if tk.exact_type == 3:
+                lineno, _ = tk.end
+                break
+        else:
+            lineno = 0
+    else:
+        # this block can be removed when python 3.6 support is dropped
+        docstring_node = node.body[0]
+        docstring = docstring_node.value.s
+        # python2.7: Code was read in bytes needs decoding to utf-8
+        # unless future unicode_literals is imported in source which
+        # make ast output unicode strings
+        if hasattr(docstring, 'decode') and not isinstance(docstring, unicode):
+            docstring = docstring.decode('utf-8')
+        lineno = docstring_node.lineno  # The last line of the string.
+
+    # This get the content of the file after the docstring last line
+    # Note: 'maxsplit' argument is not a keyword argument in python2
+    rest = '\n'.join(content.split('\n')[lineno:])
+    lineno += 1
     return docstring, rest, lineno
 
 

--- a/sphinx_gallery/tests/test_py_source_parser.py
+++ b/sphinx_gallery/tests/test_py_source_parser.py
@@ -9,11 +9,19 @@ Test source parser
 # License: 3-clause BSD
 
 from __future__ import division, absolute_import, print_function
+
+import os.path as op
+import pytest
 import sphinx_gallery.py_source_parser as sg
 
 
-def test_get_docstring_and_rest(unicode_sample):
-
+def test_get_docstring_and_rest(unicode_sample, tmpdir):
     docstring, rest, lineno = sg.get_docstring_and_rest(unicode_sample)
     assert u'Únicode' in docstring
     assert u'heiß' in rest
+    # degenerate
+    fname = op.join(str(tmpdir), 'temp')
+    with open(fname, 'w') as fid:
+        fid.write('print("hello")\n')
+    with pytest.raises(ValueError, match='Could not find docstring'):
+        sg.get_docstring_and_rest(fname)

